### PR TITLE
Outer node: compare `pos-network-node` and `substrate-node-template` implementation

### PR DESCRIPTION
**Pull request by AndreiNavoichyk**
The goal for this PR is to compare outer node implementation in `pos-network-node` and `substrate-node-template`.

### CODE REVIEW

File: node/src/cli.rs [hunk 1]
```diff
@@ -1,12 +1,31 @@
...
 #[derive(Debug, clap::Parser)]
 pub struct Cli {
+	/// Possible subcommand with parameters.
 	#[clap(subcommand)]
 	pub subcommand: Option<Subcommand>,
 
+	#[allow(missing_docs)]
 	#[clap(flatten)]
-	pub run: RunCmd,
+	pub run: sc_cli::RunCmd,
```
Comment (AndreiNavoichyk) [inline node/src/cli.rs:28]: Syntax change

File: node/src/cli.rs [hunk 2]
```diff
@@ -19,12 +38,42 @@ pub struct Cli {
...
 
+/// Possible subcommands of the main binary.
 #[derive(Debug, clap::Subcommand)]
 pub enum Subcommand {
+	/// The custom inspect subcommmand for decoding blocks and extrinsics.
+	#[clap(
+		name = "inspect",
+		about = "Decode given block or extrinsic using current native runtime."
+	)]
+	Inspect(node_inspect::cli::InspectCmd),
```
Comment (AndreiNavoichyk) [inline node/src/cli.rs:49]: Not included to default template

File: node/src/cli.rs [hunk 3]
```diff
@@ -19,12 +38,42 @@ pub struct Cli {
...
+
+	/// Sub-commands concerned with benchmarking.
+	/// The pallet benchmarking moved to the `pallet` sub-command.
+	#[clap(subcommand)]
+	Benchmark(frame_benchmarking_cli::BenchmarkCmd),
```
Comment (AndreiNavoichyk) [inline node/src/cli.rs:54]: Line change

File: node/src/cli.rs [hunk 4]
```diff
@@ -19,12 +38,42 @@ pub struct Cli {
...
+
+	/// Try some command against runtime state.
+	#[cfg(feature = "try-runtime")]
+	TryRuntime(try_runtime_cli::TryRuntimeCmd),
+
+	/// Try some command against runtime state. Note: `try-runtime` feature must be enabled.
+	#[cfg(not(feature = "try-runtime"))]
+	TryRuntime,
```
Comment (AndreiNavoichyk) [inline node/src/cli.rs:62]: Line change

File: node/src/cli.rs [hunk 5]
```diff
@@ -19,12 +38,42 @@ pub struct Cli {
...
+	Inspect(node_inspect::cli::InspectCmd),
+
+	/// Sub-commands concerned with benchmarking.
+	/// The pallet benchmarking moved to the `pallet` sub-command.
+	#[clap(subcommand)]
+	Benchmark(frame_benchmarking_cli::BenchmarkCmd),
+
+	/// Try some command against runtime state.
+	#[cfg(feature = "try-runtime")]
+	TryRuntime(try_runtime_cli::TryRuntimeCmd),
```
Comment (AndreiNavoichyk) [inline node/src/cli.rs:58]: Line change

File: node/src/cli.rs [hunk 6]
```diff
@@ -19,12 +38,42 @@ pub struct Cli {
...
+	/// Try some command against runtime state. Note: `try-runtime` feature must be enabled.
+	#[cfg(not(feature = "try-runtime"))]
+	TryRuntime,
+
 	/// Key management cli utilities
 	#[clap(subcommand)]
 	Key(sc_cli::KeySubcommand),
 
+	/// Verify a signature for a message, provided on STDIN, with a given (public or secret) key.
+	Verify(sc_cli::VerifyCmd),
```
Comment (AndreiNavoichyk) [inline node/src/cli.rs:69]: Not included to default template

File: node/src/cli.rs [hunk 7]
```diff
@@ -19,12 +38,42 @@ pub struct Cli {
...
+
+	/// Generate a seed that provides a vanity address.
+	Vanity(sc_cli::VanityCmd),
```
Comment (AndreiNavoichyk) [inline node/src/cli.rs:72]: Not included to default template

File: node/src/cli.rs [hunk 8]
```diff
@@ -19,12 +38,42 @@ pub struct Cli {
...
+
+	/// Sign a message, with a given (secret) key.
+	Sign(sc_cli::SignCmd),
```
Comment (AndreiNavoichyk) [inline node/src/cli.rs:75]: Not included to default template

File: node/src/command.rs [hunk 9]
```diff
@@ -28,33 +49,96 @@ impl SubstrateCli for Cli {
 	}
 
 	fn support_url() -> String {
-		"support.anonymous.an".into()
```
Comment (AndreiNavoichyk) [inline node/src/command.rs:31]: @yahortsaryk 
Update url

File: node/src/command.rs [hunk 10]
```diff
@@ -28,33 +49,96 @@ impl SubstrateCli for Cli {
...
+		"https://github.com/paritytech/substrate/issues/new".into()
 	}
 
 	fn copyright_start_year() -> i32 {
 		2017
 	}
 
-	fn load_spec(&self, id: &str) -> Result<Box<dyn sc_service::ChainSpec>, String> {
```
Comment (AndreiNavoichyk) [inline node/src/command.rs:38]: @AndreiNavoichyk 
Will add chain specs

File: node/src/command.rs [hunk 11]
```diff
@@ -28,33 +49,96 @@ impl SubstrateCli for Cli {
...
+						.into(),
+				),
+			"dev" => Box::new(chain_spec::development_config()),
+			"local" => Box::new(chain_spec::local_testnet_config()),
+			"fir" | "flaming-fir" => Box::new(chain_spec::flaming_fir_config()?),
+			"staging" => Box::new(chain_spec::staging_testnet_config()),
 			path =>
 				Box::new(chain_spec::ChainSpec::from_json_file(std::path::PathBuf::from(path))?),
-		})
+		};
```
Comment (AndreiNavoichyk) [inline node/src/command.rs:72]: Syntax change

File: node/src/command.rs [hunk 12]
```diff
@@ -63,29 +147,29 @@ pub fn run() -> sc_cli::Result<()> {
...
-				let PartialComponents { client, task_manager, .. } = service::new_partial(&config)?;
+				let PartialComponents { client, task_manager, .. } = new_partial(&config)?;
 				Ok((cmd.run(client, config.database), task_manager))
 			})
 		},
 		Some(Subcommand::ExportState(cmd)) => {
 			let runner = cli.create_runner(cmd)?;
 			runner.async_run(|config| {
-				let PartialComponents { client, task_manager, .. } = service::new_partial(&config)?;
+				let PartialComponents { client, task_manager, .. } = new_partial(&config)?;
```
Comment (AndreiNavoichyk) [inline node/src/command.rs:164]: Syntax change

File: node/src/service.rs [hunk 13]
```diff
@@ -1,77 +1,151 @@
...
 
-pub(crate) type FullClient =
+/// The full client type definition.
+pub type FullClient =
 	sc_service::TFullClient<Block, RuntimeApi, NativeElseWasmExecutor<ExecutorDispatch>>;
 type FullBackend = sc_service::TFullBackend<Block>;
 type FullSelectChain = sc_consensus::LongestChain<FullBackend, Block>;
 type FullGrandpaBlockImport =
-  sc_finality_grandpa::GrandpaBlockImport<FullBackend, Block, FullClient, FullSelectChain>;
+	grandpa::GrandpaBlockImport<FullBackend, Block, FullClient, FullSelectChain>;
```
Comment (AndreiNavoichyk) [inline node/src/service.rs:46]: Syntax change

File: node/src/service.rs [hunk 14]
```diff
@@ -1,77 +1,151 @@
...
 
+/// The transaction pool type defintion.
 pub type TransactionPool = sc_transaction_pool::FullPool<Block, FullClient>;
 
+/// Fetch the nonce of the given `account` from the chain state.
+///
+/// Note: Should only be used for tests.
+pub fn fetch_nonce(client: &FullClient, account: sp_core::sr25519::Pair) -> u32 {
```
Comment (AndreiNavoichyk) [inline node/src/service.rs:54]: @yahortsaryk
Don't we need this code?

File: node/src/service.rs [hunk 15]
```diff
@@ -1,77 +1,151 @@
...
+		.expect("Fetching account nonce works; qed")
+}
+
+/// Create a transaction using the given `call`.
+///
+/// The transaction will be signed by `sender`. If `nonce` is `None` it will be fetched from the
+/// state of the best block.
+///
+/// Note: Should only be used for tests.
+pub fn create_extrinsic(
```
Comment (AndreiNavoichyk) [inline node/src/service.rs:68]: Moved to `command_helper.rs`

File: node/src/service.rs [hunk 16]
```diff
@@ -1,77 +1,151 @@
...
-ServiceError,
+	sc_service::PartialComponents<
+		FullClient,
+		FullBackend,
+		FullSelectChain,
+		sc_consensus::DefaultImportQueue<Block, FullClient>,
+		sc_transaction_pool::FullPool<Block, FullClient>,
+		(
+			impl Fn(
+				node_rpc::DenyUnsafe,
```
Comment (AndreiNavoichyk) [inline node/src/service.rs:135]: Syntax change

File: node/src/service.rs [hunk 17]
```diff
@@ -1,77 +1,151 @@
...
+				sc_rpc::SubscriptionTaskExecutor,
+			) -> Result<jsonrpsee::RpcModule<()>, sc_service::Error>,
+			(
+				sc_consensus_babe::BabeBlockImport<Block, FullClient, FullGrandpaBlockImport>,
+				grandpa::LinkHalf<Block, FullClient, FullSelectChain>,
```
Comment (AndreiNavoichyk) [inline node/src/service.rs:140]: Syntax change

File: node/src/service.rs [hunk 18]
```diff
@@ -1,77 +1,151 @@
...
+				sc_consensus_babe::BabeLink<Block>,
+			),
+			grandpa::SharedVoterState,
```
Comment (AndreiNavoichyk) [inline node/src/service.rs:143]: Syntax change

File: node/src/service.rs [hunk 19]
```diff
@@ -113,7 +187,7 @@ ServiceError,
 		client.clone(),
 	);
 
-	let (grandpa_block_import, grandpa_link) = sc_finality_grandpa::block_import(
+	let (grandpa_block_import, grandpa_link) = grandpa::block_import(
```
Comment (AndreiNavoichyk) [inline node/src/service.rs:190]: Syntax change

File: node/src/service.rs [hunk 20]
```diff
@@ -161,10 +235,10 @@ ServiceError,
 
 		let justification_stream = grandpa_link.justification_stream();
 		let shared_authority_set = grandpa_link.shared_authority_set().clone();
-		let shared_voter_state = sc_finality_grandpa::SharedVoterState::empty();
+		let shared_voter_state = grandpa::SharedVoterState::empty();
```
Comment (AndreiNavoichyk) [inline node/src/service.rs:238]: Syntax change

File: node/src/service.rs [hunk 21]
```diff
@@ -161,10 +235,10 @@ ServiceError,
...
 		let shared_voter_state2 = shared_voter_state.clone();
 
-		let finality_proof_provider = sc_finality_grandpa::FinalityProofProvider::new_for_service(
+		let finality_proof_provider = grandpa::FinalityProofProvider::new_for_service(
```
Comment (AndreiNavoichyk) [inline node/src/service.rs:241]: Syntax change

File: node/src/service.rs [hunk 22]
```diff
@@ -180,18 +254,18 @@ ServiceError,
 
 		let rpc_backend = backend.clone();
 		let rpc_extensions_builder = move |deny_unsafe, subscription_executor| {
-			let deps = crate::rpc::FullDeps {
+			let deps = node_rpc::FullDeps {
```
Comment (AndreiNavoichyk) [inline node/src/service.rs:257]: Syntax change

File: node/src/service.rs [hunk 23]
```diff
@@ -180,18 +254,18 @@ ServiceError,
...
 				client: client.clone(),
 				pool: pool.clone(),
 				select_chain: select_chain.clone(),
 				chain_spec: chain_spec.cloned_box(),
 				deny_unsafe,
-				babe: crate::rpc::BabeDeps {
+				babe: node_rpc::BabeDeps {
```
Comment (AndreiNavoichyk) [inline node/src/service.rs:263]: Syntax change

File: node/src/service.rs [hunk 24]
```diff
@@ -180,18 +254,18 @@ ServiceError,
...
 					babe_config: babe_config.clone(),
 					shared_epoch_changes: shared_epoch_changes.clone(),
 					keystore: keystore.clone(),
 				},
-				grandpa: crate::rpc::GrandpaDeps {
+				grandpa: node_rpc::GrandpaDeps {
```
Comment (AndreiNavoichyk) [inline node/src/service.rs:268]: Syntax change

File: node/src/service.rs [hunk 25]
```diff
@@ -200,7 +274,7 @@ ServiceError,
 				},
 			};
 
-      crate::rpc::create_full(deps, rpc_backend.clone()).map_err(Into::into)
+			node_rpc::create_full(deps, rpc_backend.clone()).map_err(Into::into)
```
Comment (AndreiNavoichyk) [inline node/src/service.rs:277]: Syntax change

File: node/src/service.rs [hunk 26]
```diff
@@ -218,16 +292,6 @@ ServiceError,
 	})
 }
 
-/// Builds a new service for a full client.
-pub fn new_full(
```
Comment (AndreiNavoichyk) [inline node/src/service.rs:222]: Line change

File: node/src/service.rs [hunk 27]
```diff
@@ -445,7 +510,7 @@ pub fn new_full_base(
 	let keystore =
 		if role.is_authority() { Some(keystore_container.sync_keystore()) } else { None };
 
-	let config = sc_finality_grandpa::Config {
+	let config = grandpa::Config {
```
Comment (AndreiNavoichyk) [inline node/src/service.rs:513]: Syntax change

File: node/src/service.rs [hunk 28]
```diff
@@ -464,12 +529,12 @@ pub fn new_full_base(
 		// and vote data availability than the observer. The observer has not
 		// been tested extensively yet and having most nodes in a network run it
 		// could lead to finality stalls.
-		let grandpa_config = sc_finality_grandpa::GrandpaParams {
+		let grandpa_config = grandpa::GrandpaParams {
```
Comment (AndreiNavoichyk) [inline node/src/service.rs:532]: Syntax change

File: node/src/service.rs [hunk 29]
```diff
@@ -464,12 +529,12 @@ pub fn new_full_base(
...
 			config,
 			link: grandpa_link,
 			network: network.clone(),
 			telemetry: telemetry.as_ref().map(|x| x.handle()),
-			voting_rule: sc_finality_grandpa::VotingRulesBuilder::default().build(),
+			voting_rule: grandpa::VotingRulesBuilder::default().build(),
```
Comment (AndreiNavoichyk) [inline node/src/service.rs:537]: Syntax change

File: node/src/service.rs [hunk 30]
```diff
@@ -479,10 +544,274 @@ pub fn new_full_base(
 		task_manager.spawn_essential_handle().spawn_blocking(
 			"grandpa-voter",
 			None,
-			sc_finality_grandpa::run_grandpa_voter(grandpa_config)?,
+			grandpa::run_grandpa_voter(grandpa_config)?,
```
Comment (AndreiNavoichyk) [inline node/src/service.rs:547]: Syntax change

File: node/src/service.rs [hunk 31]
```diff
@@ -479,10 +544,274 @@ pub fn new_full_base(
...
+pub fn new_full(
+	config: Configuration,
+	disable_hardware_benchmarks: bool,
+) -> Result<TaskManager, ServiceError> {
+	new_full_base(config, disable_hardware_benchmarks, |_, _| ())
+		.map(|NewFullBase { task_manager, .. }| task_manager)
+}
+
+#[cfg(test)]
+mod tests {
```
Comment (AndreiNavoichyk) [inline node/src/service.rs:565]: @yahortsaryk 
Don't we need tests?

File: node/src/chain_spec.rs [hunk 32]
```diff
@@ -33,27 +61,162 @@ pub struct Extensions {
...
 
-/// Specialized `ChainSpec`. This is a specialization of the general Substrate ChainSpec type.
+/// Specialized `ChainSpec`.
 pub type ChainSpec = sc_service::GenericChainSpec<GenesisConfig, Extensions>;
+/// Flaming Fir testnet generator
+pub fn flaming_fir_config() -> Result<ChainSpec, String> {
+	ChainSpec::from_json_bytes(&include_bytes!("../res/flaming-fir.json")[..])
+}
+
+fn session_keys(
```
Comment (AndreiNavoichyk) [inline node/src/chain_spec.rs:71]: Line change

File: node/src/chain_spec.rs [hunk 33]
```diff
@@ -33,27 +61,162 @@ pub struct Extensions {
 	pub light_sync_state: sc_sync_state_rpc::LightSyncStateExtension,
 }
 
-/// Specialized `ChainSpec`. This is a specialization of the general Substrate ChainSpec type.
+/// Specialized `ChainSpec`.
 pub type ChainSpec = sc_service::GenericChainSpec<GenesisConfig, Extensions>;
+/// Flaming Fir testnet generator
+pub fn flaming_fir_config() -> Result<ChainSpec, String> {
```
Comment (AndreiNavoichyk) [inline node/src/chain_spec.rs:67]: @yahortsaryk 
Don't we need this code?

File: node/src/chain_spec.rs [hunk 34]
```diff
@@ -315,3 +362,134 @@ fn testnet_genesis(
...
+	testnet_genesis(
+		vec![authority_keys_from_seed("Alice")],
+		vec![],
+		get_account_id_from_seed::<sr25519::Public>("Alice"),
+		None,
+	)
+}
+
+/// Development config (single validator Alice)
+pub fn development_config() -> ChainSpec {
```
Comment (AndreiNavoichyk) [inline node/src/chain_spec.rs:376]: Not included to default template

File: node/src/chain_spec.rs [hunk 35]
```diff
@@ -315,3 +362,134 @@ fn testnet_genesis(
...
+		vec![],
+		None,
+		None,
+		None,
+		None,
+		Default::default(),
+	)
+}
+
+fn local_testnet_genesis() -> GenesisConfig {
```
Comment (AndreiNavoichyk) [inline node/src/chain_spec.rs:391]: Not included to default template

File: node/src/chain_spec.rs [hunk 36]
```diff
@@ -315,3 +362,134 @@ fn testnet_genesis(
...
+	testnet_genesis(
+		vec![authority_keys_from_seed("Alice"), authority_keys_from_seed("Bob")],
+		vec![],
+		get_account_id_from_seed::<sr25519::Public>("Alice"),
+		None,
+	)
+}
+
+/// Local testnet config (multivalidator Alice + Bob)
+pub fn local_testnet_config() -> ChainSpec {
```
Comment (AndreiNavoichyk) [inline node/src/chain_spec.rs:401]: Not included to default template

File: node/src/chain_spec.rs [hunk 37]
```diff
@@ -315,3 +362,134 @@ fn testnet_genesis(
...
+		None,
+		None,
+		None,
+		None,
+		Default::default(),
+	)
+}
+
+#[cfg(test)]
+pub(crate) mod tests {
```
Comment (AndreiNavoichyk) [inline node/src/chain_spec.rs:417]: @yahortsaryk 
Don't we need tests?

File: node/src/rpc.rs [hunk 38]
```diff
@@ -1,9 +1,35 @@
...
+//! It means that `client/rpc` can't have any methods that
+//! need some strong assumptions about the particular runtime.
+//!
+//! The RPCs available in this crate however can make some assumptions
+//! about how the runtime is constructed and what FRAME pallets
+//! are part of it. Therefore all node-runtime-specific RPCs can
+//! be placed here or imported from corresponding FRAME RPC definitions.
 
 #![warn(missing_docs)]
+#![warn(unused_crate_dependencies)]
```
Comment (AndreiNavoichyk) [inline node/src/rpc.rs:32]: Not included to default template

File: node/src/service.rs (hunk 31 above)
Comment (yahortsaryk) [reply node/src/service.rs:565]: Most of the tests in the old repository that are not related to runtime custom pallets, are a part of the substrate framework. They are supposed to test substrate units and they are not relevant for a code that is a client of substrate framework code.
There is no similar unit tests section in the [default substrate template](https://github.com/substrate-developer-hub/substrate-node-template/blob/polkadot-v0.9.22/node/src/service.rs).

Moreover, [all tests in the section](https://github.com/Cerebellum-Network/pos-network-node/blob/dev-cere/bin/node/cli/src/service.rs#L599) that you are referring are marked with the `#[ignore]` macro. So, I don't see a reason to keep this tests in the new repo

File: node/src/chain_spec.rs (hunk 33 above)
Comment (yahortsaryk) [reply node/src/chain_spec.rs:67]: The `flaming fir` is a name of some sandbox\testnet environment that is probably used for testing substrate features for upcoming releases. This code loads the corresponding raw chain specification from [here](https://github.com/Cerebellum-Network/pos-network-node/blob/dev-cere/bin/node/cli/res/flaming-fir.json).

I don't think we need this config file as we have our own testnet environment

File: node/src/chain_spec.rs [hunk 39]
```diff
@@ -67,129 +230,20 @@ pub fn authority_keys_from_seed(
...
 		GrandpaId,
 		BabeId,
 		ImOnlineId,
 		AuthorityDiscoveryId,
 	)>,
-  initial_nominators: Vec<AccountId>,
+	initial_nominators: Vec<AccountId>,
 	root_key: AccountId,
 	endowed_accounts: Option<Vec<AccountId>>,
-	_enable_println: bool,
```
[the remark this answers is not in the document]
Comment (yahortsaryk) [reply node/src/chain_spec.rs:190]: We probably don't need it. This variable was introduced in the [default substrate template](https://github.com/substrate-developer-hub/substrate-node-template/blob/polkadot-v0.9.22/node/src/chain_spec.rs#L133), but I don't see any usages of it. It has the `_` prefix which intentionally suppresses "var is not used" warnings.

I don't see a reason to keep it, so I'll remove it from the function declaration

File: node/src/chain_spec.rs (hunk 37 above)
Comment (yahortsaryk) [reply node/src/chain_spec.rs:417]: It looks like the code from the tests section that you are referring to, is mostly used for other tests within the old repository. There are `integration_test_config_with_two_authorities` and `integration_test_config_with_single_authority` that are declared as `pub` (public). The tests section itself is also declared as `pub(crate)`. I see three usages of those functions in other tests ([1](https://github.com/Cerebellum-Network/pos-network-node/blob/dev-cere/bin/node/cli/src/service.rs#L611), [2](https://github.com/Cerebellum-Network/pos-network-node/blob/dev-cere/bin/node/cli/src/service.rs#L803), [3](https://github.com/Cerebellum-Network/pos-network-node/blob/dev-cere/bin/node/cli/src/chain_spec.rs#L469) ), but all of them are marked with `#[ignore]` macro.

So, this test section seems redundant for us. If we need tests for chain spec configuration, we need to add them on our own specifically for our environments.

File: node/src/command.rs [hunk 40]
```diff
@@ -96,56 +180,15 @@ pub fn run() -> sc_cli::Result<()> {
 		Some(Subcommand::Revert(cmd)) => {
 			let runner = cli.create_runner(cmd)?;
 			runner.async_run(|config| {
-				let PartialComponents { client, task_manager, backend, .. } =
-					service::new_partial(&config)?;
-				let aux_revert = Box::new(|client, _, blocks| {
```
[the remark this answers is not in the document]
Comment (yahortsaryk) [reply node/src/command.rs:101]: Fixed

File: node/src/command.rs (hunk 9 above)
Comment (yahortsaryk) [reply node/src/command.rs:31]: Please, let me know what url we need to use as a link to technical support. In the old repository, there is a link to create an issue in the official substrate repository (i.e. https://github.com/paritytech/substrate/issues/new). 
Do we have a similar page for Cere blockchain node?

Comment (yahortsaryk) [reply node/src/command.rs:31]: Based on Chris Ward answer, I used https://cere.network/discord

**Pull request closed without merging**